grr/server/grr_response_server/gui/ui/components/flow_args_form/osquery_query_helper/osquery_query_helper.ts

```diff
@@ -0,0 +1,320 @@
+import {Component, ViewEncapsulation, OnDestroy} from '@angular/core';
+import {FormControl} from '@angular/forms';
+
+import {Observable, Subject, ReplaySubject} from 'rxjs';
+import {debounceTime, startWith, map, filter, distinctUntilChanged, takeUntil} from 'rxjs/operators';
+
+import {allTableSpecs, OsqueryTableSpec, nameToTable} from './osquery_table_specs';
+import {isNonNull} from '@app/lib/preconditions';
+import {SingleElementFuzzyMatcher, Match} from '@app/lib/fuzzy_matcher';
+import {MatchResultForTable} from './table_info_item/table_info_item';
+
+/** Provides functionality for composing SQL queries from Osquery table spces */
+export class QueryComposer {
+  static constructFromClause(table: OsqueryTableSpec): string {
+    return `FROM ${table.name}`
+  }
+
+  static constructSelectClause(table: OsqueryTableSpec): string {
+    const allColumnNames = table.columns.map(column => column.name);
+    const columnsWithSeparator = allColumnNames.join(',\n\t');
+    return `SELECT\n\t${columnsWithSeparator}`;
+  }
+
+  static constructWhereClause(table: OsqueryTableSpec): string {
+    const requiredColumnNames = table.columns
+      .filter(column => column.required)
+      .map(column => column.name);
+    if (requiredColumnNames.length === 0) {
+      return '';
+    }
+
+    const comment = 'TODO: Set required field as appropriate.'
+    const columnConstraints = requiredColumnNames.map(
+        colName => `${colName} LIKE "" -- ${comment}`);
+    const whereClauseArgs = columnConstraints.join('\nAND\n\t');
+
+    return `WHERE\n\t${whereClauseArgs}`;
+  }
+
+  static constructSelectAllFromTable(table: OsqueryTableSpec): string {
+    const selectClause = QueryComposer.constructSelectClause(table);
+    const fromClause = QueryComposer.constructFromClause(table);
+    const whereClause = QueryComposer.constructWhereClause(table);
+
+    return [selectClause, fromClause, whereClause, ';']
+      .filter(clause => clause !== '')
+      .join('\n');
+  }
+}
+
+/**
+ * Data structure which holds a table category name and the Osquery tables specs inside.
+ */
+class TableCategory {
+  private constructor(
+      public name: string,
+      public tableSpecs: ReadonlyArray<OsqueryTableSpec>,
+  ) { }
+
+  static fromTableSpecs(
+    name: string,
+    tableSpecs: ReadonlyArray<OsqueryTableSpec>,
+  ): TableCategory {
+    return new TableCategory(name, tableSpecs);
+  }
+
+  static fromTableNames(
+    categoryName: string,
+    tableNames: ReadonlyArray<string>,
+  ): TableCategory {
+    const tableSpecs = tableNames.map(tableName => {
+      const tableSpec = nameToTable(tableName);
+      if (!isNonNull(tableSpec)) {
+        throw Error(`Table not found: ${tableName} (category ${categoryName})`);
+      }
+      return tableSpec;
+    });
+
+    return new TableCategory(categoryName, tableSpecs);
+  }
+}
+
+declare interface ValueWithMatcher {
+  readonly value: string;
+  readonly matcher: SingleElementFuzzyMatcher;
```

Comment:
Actually this particular thing, apart from making it ineffective, is also what over-complicates the code in my opinion too. I thought there's no way to have 1 matcher and map back matches to table names, descriptions and columns, because `match(1)` returns match results not in the order that subjects were passed (and doesn't return match results for subjects which were didn't have at least 1):

I thought the dictionary approach wouldn't work if there are repeating values (e.g. column names can easily repeat), because we wouldn't be able to uniquely map matches to original subjects. Thinking of it now, if two subjects are equal, they will have the same matches, so there's no issue at all?